### DISCUSSION TITLE
BTCPay Server Greenfield Integrations

### DISCUSSION THREAD

**Discussion by Kukks** [btcpayserver/btcpayserver#2661] [Plugins & integrations]
After many years of planning and discussion, the GreenField API is now in a great state. It is stable, feature-packed, and easy to use. In terms of comparison to the BitPay compatible API that is also offered in parallel, there are no missing features.

Most of our current integrations are forks from Bitpay integrations (be it official or community-based) and so do not use the new GreenField API.  These generally consist of very old code that has been patched over and over and do not contain the best practices. Beyond that, they are also built without specific goals or feature guidelines, and thus vary greatly in what they offer and how.

We will need to create new base libraries for greenfield so that we allow developers to build their own integrations without everyone repeating the same base code needed to communicate and authenticate with the API.

Current Greenfield libraries:
* C# - https://www.nuget.org/packages/BTCPayServer.Client/
* PHP - experimental and probably prone to be replaced - https://github.com/btcpayserver/btcpayserver-greenfield-php

Current GreenField integrations:
* Shopify  (not really but it does not need to be rewritten for greenfield)

The priority for integrations:
* Magento2 -  in progress - @woutersamaey
* WooCommerce - planned to done by @ndeet
* Zapier- in progress - https://github.com/btcpayserver/zapier (beta) - @woutersamaey
* PrestaShop - (currently only Bitpay API, maybe @BitcoinMitchell can do this?) 

Others to consider:
* Ecwid (similar integration style to Shopify)
* BigCommerce (similar integration style to Shopify)
* OpenMage (magento fork) - not sure what this requires? @woutersamaey 
* WHMCS
* NopCommerce
* SimplCommerce
* GrandNode
* Drupal Commerce
* OpenCart

Impossible integrations (do not offer APIs to integrate)
* BigCartel
* Weebly
* Wix
* SiteBuilder
* SquareSpace

The majority of these integrations are in PHP, so we need to develop the PHP library to be as easy and powerful as the C# library (which is developed and used within the BTCPayServer project itself so is always the best reference implementation).

As our API is fully documented via OpenAPI v3 (aka Swagger), we can theoretically generate clients on demand, however, these are usually quite ugly and less than ideal. In my opinion, we should build the PHP library manually while offering some form of tooling to generate other libraries. 

We must also establish a core feature-set that we should strive for every integration.  Below is my initial list:
* If the integration is of a self-hosted solution, we SHOULD offer a way to spin it up in our docker deployment solution.
* Payment methods within BTCPay Server should be possible to be treated as individual payment methods in the integration.
* If the integration is of a self-hosted solution, we should offer a way to connect to BTCPay Server via Tor.
* Modal support
* Embedded mode (show a custom checkout instead of loading the BTCPay UI). This will allow integrations that need to connect to BTCPay via Tor NOT require the customer to browse via Tor
* Use the [API Key Authorization Request](https://docs.btcpayserver.org/API/Greenfield/v1/#tag/Authorization/paths/~1api-keys~1authorize/get) screen to connect an integration to a store.
* Allow refunds via BTCPay Server Pull Payments and Payouts

Please share your thoughts on all of this. This is all based on mostly my personal research and collected feedback over the past few years. We have a chance to make a superior experience for all users.

**Comment by woutersamaey** [comment]
@Kukks actually I still have some issues with Greenfield like documentation errors and missing FAQ. I hope to get some time to talk to one of you guys at some time in the future. Also have been in contact with a merchant who has some challenges that we should talk about.

I can comment on Magento:
- Greenfield for Magento 2 is 80-90% finished, but stuck for now (pending discussion)
- Greenfield for Magento 1 (now renamed to OpenMage) is at 0%. The old module still uses BitPay code, but I'll develop this part modeled after what we do in M2.

I'll take care of everything Magento.

Some personal thoughts with regard to PHP:
- Lots of frameworks now use composer, but not all. We should support both. I.e. if it's not with composer, you should be able to drop a dir in the module and get it up and running easily.
- Is auto-generating the PHP code a good idea? I have some doubts, especially since I'm still finding doc errors and issues. The generated code will reflect the errors. By writing the code manually and dogfooding, we'll encourter the real-world issues.

Your core feature list sounds good, but kindly elaborate on:
- Modal support => You mean the customer can pay on the page itself though a modal/pop-up?
- Embedded mode => You mean to proxy the request somehow over clearnet to Tor somehow? I'll need some clarification here ;-)

Payment methods within BTCPay Server should be possible to be treated as individual payment methods in the integration. => This has pros and cons so we should let the merchant decide which way to go. I say support both.

Pro = The merchant can give a discount if paid over LN for example.
Con = The merchant may need to adapt his UI and allow the customer to back and forth as he may want to use a different payment method. This results in more invoices where only 1 will be paid. Over the years's I've seen many merchants struggle with this.

**Reply by Kukks** [reply]
>Payment methods within BTCPay Server should be possible to be treated as individual payment methods in the integration. => This has pros and cons so we should let the merchant decide which way to go. I say support both.

Agreed. Will clarify in my post.

**Reply by Kukks** [reply]
>@Kukks actually I still have some issues with Greenfield like documentation errors and missing FAQ. I hope to get some time to talk to one of you guys at some time in the future. Also have been in contact with a merchant who has some challenges that we should talk about.

Yes, this is mostly due to us writing the swagger definition by hand. Let's talk!

**Reply by Kukks** [reply]
>Lots of frameworks now use composer, but not all. We should support both. I.e. if it's not with composer, you should be able to drop a dir in the module and get it up and running easily.

Agreed.

**Reply by Kukks** [reply]
>Is auto-generating the PHP code a good idea? I have some doubts, especially since I'm still finding doc errors and issues. The generated code will reflect the errors. By writing the code manually and dogfooding, we'll encourter the real-world issues.

As I said, doesn't feel that good currently and would prefer a manually written one for PHP since it is the main lib for most integrations

**Reply by Kukks** [reply]
> Modal support => You mean the customer can pay on the page itself though a modal/pop-up?

Yes, using our modal feature, which loads the invoice UI in an overlay on the same page, which usually feels smoother for customers.

**Reply by Kukks** [reply]
>Embedded mode => You mean to proxy the request somehow over clearnet to Tor somehow? I'll need some clarification here ;-)

Your use-case combines this feature + Tor support.

This feature alone simply allows bypassing the btcpay invoice UI and showing the QR code, payment address, amount, etc yourself.

**Reply by ndeet** [reply]
> > Lots of frameworks now use composer, but not all. We should support both. I.e. if it's not with composer, you should be able to drop a dir in the module and get it up and running easily.
> 
> Agreed.

Somewhat agree here but I would go composer as a default as it is a defacto standard these days. I know there are still some projects with legacy data structure like WordPress that do not support it out of the box (although there is an official way to use it as of now) but Composer is the package manager of choice like NPM for JavaScript or NuGet for C#. It is a bit like supporting PHP 5 because some outdated legacy hosts did not manage to upgrade PHP in the last 10 years. 

To sumarize:

1) Built the library with composer standard as first class citizen (best practice)

2.1) add a GitHub action or similar script to be able to download the package so you it can be used doing the default Composer convention of `require __DIR__ . '/vendor/autoload.php';` 

2.2) alternatively, maybe this is only a documentation issue because generally you can include all the classes manually using `require __DIR__.'/src/GreenfieldClient.php'; ` 

For making 2.1 and 2.2 working easily without dependency hell we should make sure to not use any external libraries. E.g. instead of using popular http client like [Guzzle](https://docs.guzzlephp.org/en/stable/) we may want to use good old PHP cURL which works well enough and is not much overhead. But I would suggest to open a separate discussion about PHP client library to discuss the details.

**Reply by ndeet** [reply]
> > Is auto-generating the PHP code a good idea? I have some doubts, especially since I'm still finding doc errors and issues. The generated code will reflect the errors. By writing the code manually and dogfooding, we'll encourter the real-world issues.
> 
> As I said, doesn't feel that good currently and would prefer a manually written one for PHP since it is the main lib for most integrations

Agree, it is less maintenance burden but the DX is a bit terrible (see [Readme](https://github.com/btcpayserver/btcpayserver-greenfield-php) e.g. using Api versions in the class names like `$result = $apiInstance->apiV1ApiKeysPost($body);` 

Also considering the last point above the auto-generated code uses Guzzle and creates hundreds of classes making manual includes cumbersome.

**Reply by ndeet** [reply]
> > Embedded mode => You mean to proxy the request somehow over clearnet to Tor somehow? I'll need some clarification here ;-)
> 
> Your use-case combines this feature + Tor support.

Tor will afaik only be possible if the webserver also supports TOR which wont be the case for 99,9% I guess. But if somebody configures TOR on their webserver they should be able to use the .onion url easily for their own backend communications in combination with the embedded mode, otherwise user needs also be able to connect via tor to btcpay which we likely can't assume.

> 
> This feature alone simply allows bypassing the btcpay invoice UI and showing the QR code, payment address, amount, etc yourself.

Nice so you can get the invoice/qr code data from api and render the QR-code directly on the page, you then also need some kind of polling the invoice every few seconds to show the user that it was payed. This is a very nice feature and will also hide the btcpayserver url (unlike redirects and javascript popup)

**Reply by ndeet** [reply]
> I can comment on Magento:
> 
> * Greenfield for Magento 2 is 80-90% finished, but stuck for now (pending discussion)

Did you already start with some PHP library or client classes for Greenfield API or did you use the generated code for this?

**Reply by woutersamaey** [reply]
> Did you already start with some PHP library or client classes for Greenfield API or did you use the generated code for this?

Actually no. I wanted the module to work without additional dependencies besides what Magento already has. I can add a library and start using that once it is done, but I see no added benefit atm. Only maintenance headaches and offering installation support to newbies who are installing the module the wrong way. I'm only using 3-4 calls, so the complexity is low and these things aren't prone to change. The previous version of the module used the PHP library and I'm glad to have everything "vanilla" now.

Magento already has Guzzle on board so that's all I'm using.

If nobody is ahead of me, I can refactor and start laying a foundation. That is definitely a way forward.

**Reply by ndeet** [reply]
Ok thanks for the info. 

> Magento already has Guzzle on board so that's all I'm using.

To avoid dependency hell and e.g. some projects depending on different major versions of Guzzle it might be a good idea to not use it all in the library as pointed out in #2663 but if there are features that Guzzle provides we may still want to use it.

> If nobody is ahead of me, I can refactor and start laying a foundation. That is definitely a way forward.

I did start playing around a bit with cURL client but wanted to discuss the requirements a bit more to not not miss anything crucial. Do you have a link to your Magento 2 code so I can take a look?

**Reply by woutersamaey** [reply]
I'm with you on #2663 . Was only using Guzzle to save time.

The code is at https://github.com/storefront-bvba/magento2-btcpay-module/tree/dev but there is no point to look it at really. I'm using JSON encode, decode and arrays. Don't have any classes and was not really going to since everything is so simple and not finding the need in this case.

I have put the coding on hold pending practical discussions on how to handle certain shopkeeper issues that some are facing. When I know the best way forward, I will finish, refactor, etc...

You'd want classes etc for when you are going for a full API library.

The only stuff that I do is:
- Creating an invoice
- Polling for invoices
- Installing a webhook
- Receiving a webhook

Nothing more. Simple calls, limited set of fields.

For the library I see the benefits as:
- Having full API coverage
- Code completion in your IDE (I didn't mind as I had the docs handy and even pointed out and fixed some doc issues)
- Not needing to work out the auth stuff (I found it easy but on Mattermost many newcomers struggle with it)
- Not needing to work out the webhook secret header signature (same)

**Comment by ndeet** [comment]
I added a PHP Greenfield library discussion here to not clutter this discussion with PHP specific implementation details, hope ok for everyone otherwise let me know https://github.com/btcpayserver/btcpayserver/discussions/2663
